tests/tests/integration_test.rs

```diff
@@ -247,78 +216,78 @@ async fn veracruz_phase3_string_edit_distance_four_clients() {
-            let mut client = veracruz_client::VeracruzClient::new(
-                cert_key_dir(DATA_CLIENT_CERT).as_path(),
-                cert_key_dir(DATA_CLIENT_KEY).as_path(),
-                &policy_json,
-            )?;
-
-            let data_filename = data_dir(LINEAR_REGRESSION_DATA);
-            info!("### data provider read input.");
-            let data = read_local_file(&data_filename).unwrap();
-            info!("### data provider send input.");
-            client.write_file("/input/linear-regression.dat", &data)?;
-            info!("### data provider read result.");
-            client.request_compute("/program/linear-regression.wasm")?;
-            client.read_file("/output/linear-regression.dat")?;
-            info!("### data provider request shutdown.");
-            client.request_shutdown()?;
-            Result::<()>::Ok(())
-        };
-
-        let result = futures::future::try_join3(
-            server_handle,
-            program_provider_handle,
-            data_provider_handle,
+#[test]
+fn veracruz_phase4_linear_regression_two_clients_parallel() {
+    let (_, policy_json, _) =
+        read_policy(policy_dir(LINEAR_REGRESSION_PARALLEL_POLICY).as_path()).unwrap();
+    let policy = Policy::from_json(&policy_json).unwrap();
+
+    proxy_attestation_setup(policy.proxy_attestation_server_url().clone());
+
+    server_tls_loop(policy_json.clone());
+
+    let policy_json_cloned = policy_json.clone();
+    let client1_handle = thread::spawn(move || -> Result<()> {
+        info!("### program provider start.");
+        let mut client = veracruz_client::VeracruzClient::new(
+            cert_key_dir(PROGRAM_CLIENT_CERT).as_path(),
+            cert_key_dir(PROGRAM_CLIENT_KEY).as_path(),
+            &policy_json_cloned,
+        )?;
+        let prog_path = program_dir(LINEAR_REGRESSION_WASM);
+        info!("### program provider read binary.");
+        let program_data = read_local_file(prog_path).unwrap();
+        info!("### program provider send binary.");
+        client.write_file("/program/linear-regression.wasm", &program_data)?;
+        Result::<()>::Ok(())
+    });
+
+    let policy_json_cloned = policy_json.clone();
+    let client2_handle = thread::spawn(move || -> Result<()> {
+        info!("### data provider start.");
+        let mut client = veracruz_client::VeracruzClient::new(
+            cert_key_dir(DATA_CLIENT_CERT).as_path(),
+            cert_key_dir(DATA_CLIENT_KEY).as_path(),
+            &policy_json_cloned,
+        )
+        .unwrap();
+
+        let data_filename = data_dir(LINEAR_REGRESSION_DATA);
+        info!("### data provider read input.");
+        let data = read_local_file(&data_filename).unwrap();
+        info!("### data provider send input.");
+        client.write_file("/input/linear-regression.dat", &data)?;
+        Result::<()>::Ok(())
+    });
+
+    client1_handle.join().unwrap().unwrap();
+    client2_handle.join().unwrap().unwrap();
```

Comment:
Per our discussion yesterday: it looks like the data client and the program provider client are parallel. However, since they are both `join`ed before the result retriever starts, they are not in parallel with the result retriever. I'd be interested in seeing what happens if the result retriever were parallel, but this is not a requirement before getting this PR through.

Reply:
I've just tried it: if the result retriever runs first then it fails: `Failed to send TLS data.  Error produced: NoEnt.`